src/lib/net/src/packets/incoming/keep_alive.rs

```diff
@@ -13,15 +15,42 @@ pub struct IncomingKeepAlivePacket {
 
 impl IncomingPacket for IncomingKeepAlivePacket {
     async fn handle(self, conn_id: usize, state: Arc<ServerState>) -> NetResult<()> {
-        let mut last_keep_alive = state.universe.get_mut::<KeepAlive>(conn_id)?;
+        // TODO handle errors.
+        let last_keep_alive = state.universe.get_mut::<OutgoingKeepAlivePacket>(conn_id)?;
+
         if self.id != last_keep_alive.id {
             debug!(
-                "Invalid keep alive packet received from {:?} with id {:?} (expected {:?})",
+                "Invalid keep alive packet received from entity {:?} with id {:?} (expected {:?})",
                 conn_id, self.id, last_keep_alive.id
             );
+            return NetResult::Err(crate::errors::NetError::Packet(
+                crate::errors::PacketError::InvalidState(0x18),
+            ));
             // TODO Kick player
+        }
+
+        let result = state.universe.get_mut::<IncomingKeepAlivePacket>(conn_id);
+
+        if result.is_err() {
+            let err = result.as_ref().err().unwrap();
+            if matches!(err, ECSError::ComponentTypeNotFound) {
+                state
+                    .universe
+                    .add_component(conn_id, IncomingKeepAlivePacket { id: self.id })?;
+                let mut last_received_keep_alive = state.universe.get_mut(conn_id)?;
+                *last_received_keep_alive = self;
+            } else {
+                warn!(
+                    "Failed to get or create <IncomingKeepAlive> component: {:?}",
+                    err
+                );
+                return Err(crate::errors::NetError::ECSError(result.err().unwrap()));
+            }
         } else {
-            *last_keep_alive = KeepAlive::from(self.id);
+            let mut last_received_keep_alive: ComponentRefMut<'_, IncomingKeepAlivePacket> =
+                result.unwrap();
+
+            *last_received_keep_alive = self;
```

Comment:
Why not just do a match statement?
```rs
match result {
    Ok(component) => {
        // ...
    },
    Err(e) => {
        // ...
    }
}
```

Reply:
doesnt fix the problem of the error?

Reply:
```rs
state.universe.get_mut::<IncomingKeepAlivePacket>(conn_id)
```
 is the culprit

Reply:
Can this be a deadlock from dashmap?

Reply:
If something stupid is being done, probably. I haven't read the code in too much depth, but this shouldn't be the case.

Reply:
it wasnt, I think the error system is a bit inconsistent, I expected only ComponentTypeNotFound, but also got ComponentRetrievalError, when they're both reporting what should be the same error since the IncomingKeepAlive hasn't been created yet, I've fixed this in #122